File: swift/Side/15_strain/Sources/Strain/Strain.swift
extension Collection {

    func keep(_ predicate: (Element) -> Bool) -> [Element] {
        return self.mapArray(predicate, shouldAppend: true)
    }

    func discard(_ predicate: (Element) -> Bool) -> [Element] {
        return self.mapArray(predicate, shouldAppend: false)
    }

    private func mapArray(_ predicate: (Element) -> Bool, shouldAppend: Bool) -> [Element] {
        // You could also do this by just looping over self and appending the element into a returnArray
        return self.compactMap {
            return (predicate($0) == shouldAppend) ? $0 : nil
        }
    }

}

File: typescript/Side/Medium/02_strain/strain.ts
export function keep<T>(array: T[], predicate: (element: T) => Boolean): T[] {
    return mapArray(array, predicate, true)
}

export function discard<T>(array: T[], predicate: (element: T) => Boolean): T[] {
    return mapArray(array, predicate, false)
}

function mapArray<T>(array: T[], predicate: (element: T) => Boolean, shouldKeep: Boolean): T[] {
    const returnArray: T[] = []

    for (const arrayElement of array) {
        if (predicate(arrayElement) === shouldKeep) {
            returnArray.push(arrayElement)
        }
    }

    return returnArray
}